### PR TITLE
Change the new datastore to use Metadata to store file information

### PR DESCRIPTION
**Pull request by romain-intel**
The new datastore would initially store some extra information in the file itself (in a header). This uses the new metadata support in S3 to store this information.

### CODE REVIEW

File: metaflow/datastore/content_addressed_store.py [hunk 1]
```diff
@@ -145,38 +138,49 @@ def load_blobs(self, keys, force_raw=False):
...
+                            info_format,
+                            r.read(struct.calcsize(info_format)))
+                        if file_magic == magic:
+                            unpack_code = getattr(
+                                self, '_unpack_v%d' % version, None)
                         else:
-                            # This is the backward compatible mode
                             r.seek(0)
+                            # This is the backward compatible mode
                             unpack_code = getattr(
```
Comment (tuulos) [inline metaflow/datastore/content_addressed_store.py:170]: why not just ```
unpack_code = self._unpack_backward_compatible
```

File: metaflow/datastore/content_addressed_store.py [hunk 2]
```diff
@@ -145,38 +138,49 @@ def load_blobs(self, keys, force_raw=False):
...
+                        version = meta.get('cas_version', -1)
+                        if version == -1:
+                            raise DataException(
+                                "Could not extract encoding version for %s" % path)
+                        unpack_code = getattr(self, '_unpack_v%d' % version, None)
+                    if unpack_code is None:
                         raise DataException(
                             "Unknown encoding version %d for %s -- the artifact "
                             "is either corrupt or you need to update Metaflow"
                             % (version, path))
```
Comment (tuulos) [inline metaflow/datastore/content_addressed_store.py:182]: `version` may be undefined here. Just print the `path`.

File: metaflow/datastore/content_addressed_store.py (hunk 1 above)
Comment (romain-intel) [reply metaflow/datastore/content_addressed_store.py:170]: Yes, dumb copy paste without thinking...

File: metaflow/datastore/content_addressed_store.py (hunk 2 above)
Comment (romain-intel) [reply metaflow/datastore/content_addressed_store.py:182]: Fair point. As discussed. We can hopefully remove this as well (or at least simplify some of it).

Comment (romain-intel) [reply metaflow/datastore/content_addressed_store.py:182]: Done: I moved this code to the else part (only relevant part after refactoring to remove the version stored directly in the file.

File: metaflow/datastore/content_addressed_store.py (hunk 1 above)
Comment (romain-intel) [reply metaflow/datastore/content_addressed_store.py:170]: Done and simplified.

**Pull request merged**